rptest/datalake: test metadata interoperability with 3rd party system

nvartolomei opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
rptest/datalake: test metadata interoperability with 3rd party system

WillemKauf commented on tests/rptest/tests/datalake/3rdparty_rewrite.py:
```diff
@@ -0,0 +1,118 @@
+                assert initial_parquet_files >= num_kafka_partitions, f"Expecting at least {num_kafka_partitions} files, got {initial_parquet_files}"
+
+                spark.run_query_fetch_all(
+                    f"CALL `redpanda-iceberg-catalog`.system.rewrite_data_files(\"redpanda.{self.topic_name}\")"
```
This line here is what actually triggers compaction (along with the trino command above).

We should have a `datalake_services::do_maintenance_work()` function (or something along those lines) for future re-use, which triggers these spark/trino queries.

nvartolomei replied:
I'd rather not generalize it until I have some experience with some additional vendors. Not sure what the method should be called and what work/how much work it should do yet. Extracted it to a separate method in the test class for now.

WillemKauf replied:
sounds good!

nvartolomei replied:
Actually, came up with a generalization. Also added a method for checking which engines support maintenance jobs. We should be able to build a grid with ticks re what is tested later 👌

nvartolomei committed:
rptest/s: add run_query_fetch_one method to query engine

nvartolomei committed:
rptest/s: query engines: count_parquet_files and optimize_parquet_files

Used for testing interoperability with 3rd party vendors. I.e. that
redpanda continues to operate after data and metadata is mutated by a
3rd party.

nvartolomei committed:
rptest/s: add DatalakeVerifier.oneshot method

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60089-->
<!--e21e7931-7902-4280-8e24-f59eb88327f8-->
<!--title-->
## Retry command for [Build#60089](https://buildkite.com/redpanda/redpanda/builds/60089)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_enable_features@{"disable_trial":false,"feature":5,"install_license":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--366b6ed8-a2a4-4954-a490-930e004eaae8-->
<!--TEST RESULTS-->
<!--6d8f03c2-d6c8-4394-a449-5b2f2ee8a176-->
<!--TEST RESULTS-->
<!--73434163-7c4e-4d6d-a7c7-d7a844c54a4f-->
<!--TEST RESULTS-->
<!--35fc2102-85a2-4c9b-9bd3-400273fb0947-->
<!--TEST RESULTS-->
<!--8b5728d0-68af-48d0-a94a-11bbef74c31c-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60089">build#60089</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60089#0193f576-ce17-455b-91e6-2f265aaa9185 | FLAKY | 2/6 |
| rptest.tests.enterprise_features_license_test.EnterpriseFeaturesTest.test_enable_features.feature=Feature.oidc.install_license=True.disable_trial=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/60089#0193f576-ce16-4642-b44b-4eb53b3c9ee5 | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60100">build#60100</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60100#0193f642-b1a2-4bf2-af5d-e55ed5e9b6f4 | FLAKY | 1/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60178">build#60178</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60178#01941785-f08b-45dd-85b4-2851b2d2bb86 | FLAKY | 2/6 |
| rptest.tests.e2e_topic_recovery_test.EndToEndTopicRecovery.test_restore_with_aborted_tx.recovery_overrides=.retention.local.target.bytes.1024.redpanda.remote.write.True.redpanda.remote.read.True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/60178#01941785-f089-4e4a-9d39-60eb4a8babe0 | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61916">build#61916</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61916#0195154c-628e-4b65-84a7-3915c9dea3ee | FLAKY | 1/4 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61916#0195154f-7d44-48b5-b5de-91d8b2d3f862 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/61916#0195154f-7d44-48b5-b5de-91d8b2d3f862 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61975">build#61975</a></summary>


| test_id                                                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61975#01951afb-bad1-44bb-ba7f-38b56df1503d | FLAKY         | 1/2      |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61975#01951ae6-a708-428b-bb72-f34c0d4232b2 | FLAKY         | 1/2      |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61975#01951ae6-a70c-40b6-a663-b6adfb3a70bd | FLAKY         | 1/2      |
</details>
<!--/body-->

andrwng commented on tests/rptest/services/spark_service.py:
```diff
@@ -149,6 +149,28 @@ def escape_identifier(self, table: str) -> str:
+    def can_run_maintenance(self):
+        return True
+
+    def run_sample_maintenance_task(self, namespace, table) -> None:
+        # Metadata query
+        # https://iceberg.apache.org/docs/1.6.1/spark-queries/#files
+        initial_parquet_files = self.run_query_fetch_one(
+            f"SELECT count(*) FROM {namespace}.{table}.files")[0]
+
+        # Want at least 2 files to be able to assert that optimization did something.
+        assert initial_parquet_files >= 2, f"Expecting at least 2 files, got {initial_parquet_files}"
+
+        # Spark Procedures provided by Iceberg SQL Extensions
+        # https://iceberg.apache.org/docs/1.6.1/spark-procedures/#rewrite_data_files
+        self.run_query_fetch_one(
+            f"CALL `redpanda-iceberg-catalog`.system.rewrite_data_files(\"{namespace}.{table}\")"
+        )
+
+        optimized_parquet_files = self.run_query_fetch_one(
+            f"SELECT count(*) FROM {namespace}.{table}.files")[0]
+        assert optimized_parquet_files < initial_parquet_files, f"Expecting fewer files after optimize, got {optimized_parquet_files}"
```
IMO these service classes should only hold operations, not validations. To that end, can we split this into some `count_parquet_files()` and `optimize_parquet_files()`? There are many different kinds of maintenance https://iceberg.apache.org/docs/1.5.1/maintenance/ and it'll be much easier to reuse if we're more descriptive and have smaller, tighter primitives.

nvartolomei replied:
@andrwng I see what you mean but here I wanted to give more freedom to these maintenance tasks implementation, i.e. it might as well be a metadata only maintenance. I documented the expectations in the code doc https://github.com/redpanda-data/redpanda/blob/9874fa01f80ff843abf586921973932b5b13bc47/tests/rptest/tests/datalake/query_engine_base.py#L70

`optimize_parquet_files` and `count_parquet_files` would work for the 2 engines we have now, do you reckon all of them expose the same primitives so that we could have a single test for all of them?

WillemKauf replied:
I agree that `run_sample_maintenance_task()` seems a bit vague, whereas the iceberg docs Andrew provided has a concrete list of tasks we could enumerate, these being:

- [ExpireSnapshots](https://iceberg.apache.org/javadoc/1.7.1/org/apache/iceberg/ExpireSnapshots.html)
- [DeleteOrphanFiles](https://iceberg.apache.org/javadoc/1.7.1/org/apache/iceberg/actions/DeleteOrphanFiles.html)
- [RewriteDataFiles](https://iceberg.apache.org/javadoc/1.7.1/org/apache/iceberg/actions/RewriteDataFiles.html)
- [RewriteManifests](https://iceberg.apache.org/javadoc/1.7.1/org/apache/iceberg/actions/RewriteManifests.html)

Maybe these are the primitives we should be working with and leave the assertions to the user.

nvartolomei replied:
> Maybe these are the primitives we should be working with and leave the assertions to the user.

Assertions like ...? I don't understand what the exact proposal is.

andrwng replied:
I believe Willem's referring to assert statements in `run_sample_maintenance_task()` -- to my earlier point, I feel strongly that they don't belong in this Spark service library. They belong in test bodies, or other test verifiers.

>here I wanted to give more freedom to these maintenance tasks implementation...do you reckon all of them expose the same primitives so that we could have a single test for all of them?

I can see how it's a nice property to have, that most query engines can implement some kind of maintenance, and so would be able to slot into the new test. To your point, I don't think every engine will always support the exact set of maintenance operations I posted/suggested.  IMO once we come across such an engine, we should decide what to do at that time, rather than preemptively introducing a loosely defined primitive -- it's hard for me to imagine this maintenance function will be used widely by test authors.

If you insist that "arbitrary maintenance" belongs here, then please also at least split out `optimize_parquet_files` and `count_parquet_files` (or your preferred names) into the Spark and Trino services and have `run_sample_maintenance_task()` call them.

nvartolomei replied:
Replaced `run_sample_maintenance_task` with `optimize_parquet_files` and `count_parquet_files`.

andrwng commented on tests/rptest/tests/datalake/3rdparty_maintenance_test.py:
```diff
@@ -0,0 +1,78 @@
+from ducktape.mark import matrix
```
nit: license header

andrwng commented on tests/rptest/tests/datalake/query_engine_base.py:
```diff
@@ -59,3 +60,22 @@ def max_translated_offset(self, namespace, table, partition) -> int:
         with self.run_query(query) as cursor:
             return cursor.fetchone()[0]
+
+    def can_run_maintenance(self) -> bool:
+        """
+        Returns True if the query engine supports running maintenance operations.
+        """
+        return False
```
nit: maybe i'm missing it, is this used or will this have immediate use (i.e. a pending PR)? If not, let's remove it until it has an immediate use

andrwng commented on tests/rptest/tests/datalake/3rdparty_maintenance_test.py:
```diff
@@ -0,0 +1,78 @@
+            self._translate_sample_data(dl)
+
+            # Run maintenance to rewrite the data.
+            dl.any_query_engine().run_sample_maintenance_task(
```
> any_query_engine()

I think it'd be a bit clearer to use `dl.service(query_engine)` -- we know what query engine we want, it's not "any". Or maybe just make this `query_engine()` and assert that exactly one is specified. It feels like we're making it just a bit harder to read/debug with an "any"

nvartolomei replied:
👍

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60178-->
<!--ac575594-2b11-4843-b075-c720579f334e-->
<!--title-->
## Retry command for [Build#60178](https://buildkite.com/redpanda/redpanda/builds/60178)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/e2e_topic_recovery_test.py::EndToEndTopicRecovery.test_restore_with_aborted_tx@{"cloud_storage_type":2,"recovery_overrides":{"redpanda.remote.read":true,"redpanda.remote.write":true,"retention.local.target.bytes":1024}}
```
<!--/body-->

dotnwat commented:
got a merge conflict

dotnwat commented on tests/rptest/tests/datalake/query_engine_base.py:
```diff
@@ -33,12 +33,9 @@ def run_query(self, query):
-                cursor.execute(query)
-                yield cursor
-            finally:
-                cursor.close()
```
was this redundant? it seems like it was handling the cursor closing, and the outer try block was handling the client closing.

dotnwat commented on tests/rptest/tests/datalake/query_engine_base.py:
```diff
@@ -60,3 +60,16 @@ def max_translated_offset(self, namespace, table, partition) -> int:
         with self.run_query(query) as cursor:
             return cursor.fetchone()[0]
+
+    def run_sample_maintenance_task(self, namespace, table) -> None:
```
i'm not sure what `_sample_` in the name refers to

andrwng commented on tests/rptest/tests/datalake/query_engine_base.py:
```diff
@@ -66,3 +66,11 @@ def max_translated_offset(self, namespace, table, partition) -> int:
+
+    @abstractmethod
+    def count_parquet_files(self, namespace, table) -> int:
+        ...
```
What does this look like when not implemented? Is it a no-op? Should we raise an exception like we do elsewhere?

nvartolomei replied:
You can't actually instantiate a class which inherits from `abc.ABC` and has any `@abstractmethod` members. The body is dead/unreachable code.

```
>>> from abc import ABC, abstractmethod
>>> class A(ABC):
...     @abstractmethod
...     def a(self):
...         print("foo")
...
>>> A().a()
Traceback (most recent call last):
  File "<python-input-13>", line 1, in <module>
    A().a()
    ~^^
TypeError: Can't instantiate abstract class A without an implementation for abstract method 'a'
>>>
```

andrwng commented on tests/rptest/tests/datalake/datalake_verifier.py:
```diff
@@ -367,3 +367,13 @@ def stop(self):
         finally:
             if self._consumer:
                 self._consumer.close()
+        assert self._max_queried_offsets == self._max_consumed_offsets, "Mismatch between maximum offsets in topic vs iceberg table"
```
is this intentionally duplicated from L362?

nvartolomei replied:
looks like bad merge conflict resolution, i'll remove it

andrwng reviewed (dismissed):
Pretty much LGTM! Thanks for the edits, just a couple questions

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-24643-v24.3.x-743 remotes/upstream/v24.3.x
git cherry-pick -x 2c931c54b3 6cd66b68b6 c35332dc0e 20e89c0ba3
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/13402138673).
